pkg/utils/report/results.go

```diff
@@ -86,72 +87,76 @@ func SeverityFromString(severity string) policyreportv1alpha2.PolicySeverity {
 	return ""
 }
 
-func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
-	pol := response.Policy()
-	var results []policyreportv1alpha2.PolicyReportResult
-	if pol.GetType() == engineapi.KyvernoPolicyType {
-		key, _ := cache.MetaNamespaceKeyFunc(pol.AsKyvernoPolicy())
-		for _, ruleResult := range response.PolicyResponse.Rules {
-			annotations := pol.GetAnnotations()
-			result := policyreportv1alpha2.PolicyReportResult{
-				Source:  kyverno.ValueKyvernoApp,
-				Policy:  key,
-				Rule:    ruleResult.Name(),
-				Message: ruleResult.Message(),
-				Result:  toPolicyResult(ruleResult.Status()),
-				Scored:  annotations[kyverno.AnnotationPolicyScored] != "false",
-				Timestamp: metav1.Timestamp{
-					Seconds: time.Now().Unix(),
-				},
-				Category: annotations[kyverno.AnnotationPolicyCategory],
-				Severity: SeverityFromString(annotations[kyverno.AnnotationPolicySeverity]),
-			}
-			if ruleResult.Exception() != nil {
-				result.Properties = map[string]string{
-					"exception": ruleResult.Exception().Name,
-				}
-			}
-			pss := ruleResult.PodSecurityChecks()
-			if pss != nil {
-				var controls []string
-				for _, check := range pss.Checks {
-					if !check.CheckResult.Allowed {
-						controls = append(controls, check.ID)
-					}
-				}
-				if len(controls) > 0 {
-					sort.Strings(controls)
-					result.Properties = map[string]string{
-						"standard": string(pss.Level),
-						"version":  pss.Version,
-						"controls": strings.Join(controls, ","),
-					}
-				}
-			}
-			if result.Result == "fail" && !result.Scored {
-				result.Result = "warn"
-			}
-			results = append(results, result)
+func addProperty(k, v string, result *policyreportv1alpha2.PolicyReportResult) {
+	if result.Properties == nil {
+		result.Properties = map[string]string{}
+	}
+
+	result.Properties[k] = v
+}
+
+func ToPolicyReportResult(policyType engineapi.PolicyType, policyName string, ruleResult engineapi.RuleResponse, annotations map[string]string, resource *corev1.ObjectReference) policyreportv1alpha2.PolicyReportResult {
+	result := policyreportv1alpha2.PolicyReportResult{
+		Source:  kyverno.ValueKyvernoApp,
+		Policy:  policyName,
+		Rule:    ruleResult.Name(),
```

Comment:
If the policy type is ValidatingAdmissionPolicy, the `rule` field should be unset.

Reply:
Will `ruleResult.Name()` not be an empty string then?

Reply:
I tried it and this seems to work:

```go
func EngineResponseToReportResults(response engineapi.EngineResponse) []policyreportv1alpha2.PolicyReportResult {
	policy := response.Policy()
	policyType := policy.GetType()
	policyName := policy.GetName()
	annotations := policy.GetAnnotations()

	var results []policyreportv1alpha2.PolicyReportResult
	for _, ruleResult := range response.PolicyResponse.Rules {
		result := ToPolicyReportResult(policyType, policyName, ruleResult, annotations, nil)
		results = append(results, result)
	}

	return results
}
```

Reply:
After this change, I was able to locally run the Chainsaw tests successfully:

```sh
chainsaw test --test-dir ./validating-admission-policy-reports --config .chainsaw.yaml
```

```sh
Version: 0.1.7
Loading config (.chainsaw.yaml)...
- Using test file: chainsaw-test.yaml
- TestDirs [./validating-admission-policy-reports]
- SkipDelete false
- FailFast true
- ReportFormat ''
- ReportName 'chainsaw-report'
- Namespace ''
- FullName true
- IncludeTestRegex ''
- ExcludeTestRegex '_.+'
- ApplyTimeout 5s
- AssertTimeout 1m30s
- CleanupTimeout 30s
- DeleteTimeout 15s
- ErrorTimeout 1m30s
- ExecTimeout 5s
- Parallel 1
- ForceTerminationGracePeriod 5s
- DelayBeforeCleanup 3s
- NoCluster false
Loading tests...
- validating-admission-policy-fail (validating-admission-policy-reports/background/validating-admission-policy-fail)
- validating-admission-policy-pass (validating-admission-policy-reports/background/validating-admission-policy-pass)
- with-validating-admission-policy-binding-fail (validating-admission-policy-reports/background/with-validating-admission-policy-binding-fail)
- with-validating-admission-policy-binding-no-reports (validating-admission-policy-reports/background/with-validating-admission-policy-binding-no-reports)
- with-validating-admission-policy-binding-pass (validating-admission-policy-reports/background/with-validating-admission-policy-binding-pass)
- events (validating-admission-policy-reports/events)
Loading values...
Running tests...
=== RUN   chainsaw
=== PAUSE chainsaw
=== CONT  chainsaw
=== RUN   chainsaw/validating-admission-policy-reports/background/validating-admission-policy-fail[validating-admission-policy-fail]
=== PAUSE chainsaw/validating-admission-policy-reports/background/validating-admission-policy-fail[validating-admission-policy-fail]
=== RUN   chainsaw/validating-admission-policy-reports/background/validating-admission-policy-pass[validating-admission-policy-pass]
=== PAUSE chainsaw/validating-admission-policy-reports/background/validating-admission-policy-pass[validating-admission-policy-pass]
=== RUN   chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-fail[with-validating-admission-policy-binding-fail]
=== PAUSE chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-fail[with-validating-admission-policy-binding-fail]
=== RUN   chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-no-reports[with-validating-admission-policy-binding-no-reports]
=== PAUSE chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-no-reports[with-validating-admission-policy-binding-no-reports]
=== RUN   chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-pass[with-validating-admission-policy-binding-pass]
=== PAUSE chainsaw/validating-admission-policy-reports/background/with-validating-admission-policy-binding-pass[with-validating-admission-policy-binding-pass]
=== RUN   chainsaw/validating-admission-policy-reports/events[events]
=== PAUSE chainsaw/validating-admission-policy-reports/events[events]
=== CONT  chainsaw/validating-admission-policy-reports/background/validating-admission-policy-fail[validating-admission-policy-fail]
    | 10:54:19 | validating-admission-policy-fail | @setup   | CREATE    | OK    | v1/Namespace @ chainsaw-choice-snipe
    | 10:54:19 | validating-admission-policy-fail | step-01  | TRY       | RUN   |
    | 10:54:19 | validating-admission-policy-fail | step-01  | APPLY     | RUN   | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | CREATE    | OK    | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | APPLY     | DONE  | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | ASSERT    | RUN   | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | ASSERT    | DONE  | apps/v1/Deployment @ chainsaw-choice-snipe/deployment-fail
    | 10:54:19 | validating-admission-policy-fail | step-01  | TRY       | DONE  |
    | 10:54:19 | validating-admission-policy-fail | step-02  | TRY       | RUN   |
    | 10:54:19 | validating-admission-policy-fail | step-02  | APPLY     | RUN   | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | CREATE    | OK    | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | APPLY     | DONE  | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | ASSERT    | RUN   | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | ASSERT    | DONE  | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:19 | validating-admission-policy-fail | step-02  | TRY       | DONE  |
    | 10:54:19 | validating-admission-policy-fail | step-03  | TRY       | RUN   |
    | 10:54:19 | validating-admission-policy-fail | step-03  | ASSERT    | RUN   | wgpolicyk8s.io/v1alpha2/PolicyReport @ chainsaw-choice-snipe/*
    | 10:54:29 | validating-admission-policy-fail | step-03  | ASSERT    | DONE  | wgpolicyk8s.io/v1alpha2/PolicyReport @ chainsaw-choice-snipe/*
    | 10:54:29 | validating-admission-policy-fail | step-03  | TRY       | DONE  |
    | 10:54:32 | validating-admission-policy-fail | @cleanup | DELETE    | RUN   | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:32 | validating-admission-policy-fail | @cleanup | DELETE    | OK    | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    | 10:54:32 | validating-admission-policy-fail | @cleanup | DELETE    | DONE  | admissionregistration.k8s.io/v1alpha1/ValidatingAdmissionPolicy @ check-deployment-replicas-02
    
    ...

Tests Summary...
- Passed  tests 6
- Failed  tests 0
- Skipped tests 0
Done.
```